sklearn/datasets/openml.py

```diff
@@ -0,0 +1,425 @@
+        version from OpenML that is annotated as active. Any other string
+        values except "active" are treated as integer.
+
+    Returns
+    -------
+    json_data['data']['dataset'][0]: json
+        json representation of the first dataset object that adhired to the
+        search criteria
+
+    """
+    if version == "active":
+        # situation in which we return the oldest active version
+        url = _SEARCH_NAME.format(name) + "/status/active/"
+        error_msg = "No active dataset {} found.".format(name)
+        json_data = _get_json_content_from_openml_api(url, error_msg, True)
+        return json_data['data']['dataset'][0]
+
+    # an integer version has been provided
+    url = (_SEARCH_NAME + "/data_version/{}").format(name, version)
+    json_data = _get_json_content_from_openml_api(url, None, False)
+    if json_data is None:
+        # we can do this in 1 function call if OpenML does not require the
+        # specification of the dataset status (i.e., return datasets with a
+        # given name / version regardless of active, deactivated, etc. )
+        # TODO: feature request OpenML.
+        url += "/status/deactivated"
+        error_msg = "Dataset {} with version {} not found.".format(name,
+                                                                   version)
+        json_data = _get_json_content_from_openml_api(url, error_msg, True)
+
+    return json_data['data']['dataset'][0]
+
+
+def _get_data_description_by_id(data_id):
+    # OpenML API function: https://www.openml.org/api_docs#!/data/get_data_id
+    url = _DATA_INFO.format(data_id)
+    error_message = "Dataset with id {} not found.".format(data_id)
+    json_data = _get_json_content_from_openml_api(url, error_message, True)
+    return json_data['data_set_description']
+
+
+def _get_data_features(data_id):
+    # OpenML function:
+    # https://www.openml.org/api_docs#!/data/get_data_features_id
+    url = _DATA_FEATURES.format(data_id)
+    error_message = "Dataset with id {} not found.".format(data_id)
+    json_data = _get_json_content_from_openml_api(url, error_message, True)
+    return json_data['data_features']['feature']
+
+
+def _download_data_arff(file_id):
+    # Accesses an ARFF file on the OpenML server. Documentation:
+    # https://www.openml.org/api_data_docs#!/data/get_download_id
+    url = _DATA_FILE.format(file_id)
+    response = urlopen(url)
+    if PY2:
+        # Because Python2.7 numpy can't handle unicode
+        arff_file = arff.loads(response.read())
+    else:
+        arff_file = arff.loads(response.read().decode('utf-8'))
+
+    response.close()
+    return arff_file
+
+
+def _convert_numericals(data, features):
+    # converts all numerical columns in the numpy array into floats
+    for feature in features:
+        if feature['data_type'] == "numeric":
+            idx = int(feature['index'])
+            data[:, idx] = data[:, idx].astype(np.float)
```

Comment:
This seems strange to me. Surely the dtype of `data` remains `object`. I don't see how this change does anything. If all features are numeric, we can convert the array altogether into float, but as long as we are using numpy arrays of simple dtype, you cannot convert one column at a time.

Reply:
The dtype of the array remains unchanged indeed, but the column will have floats (with dtype object) rather than strings (with dtype object)

Reply:
liac-arff return string dtype for numeric columns??? That seems strange behaviour to me :(

Reply:
I will check this.

Reply:
Ah, from the code I assumed it were strings (csv is all text in the end :-)), but indeed, the object array already seems to be a mixture of floats and strings.

Reply:
You are right, this function is not necessary. Without it, columns do have the correct data type as well

(this behavior is tested in the unit tests, currently on line 84 in the private test fn _fetch_dataset_from_openml).